File: src/Services/EngagementService/EngagementService.Domain/Entities/RewardTransaction.cs
namespace EngagementService.Domain.Entities;

public sealed class RewardTransaction
{
    public int TransactionId { get; set; }
    public int UserId { get; set; }
    public int? RewardId { get; set; }
    public DateTime? CreatedAt { get; set; }
    public string Type { get; set; } = null!;
    public int Points { get; set; }
    public string? Description { get; set; }
    public int? ReportId { get; set; }
    public string Status { get; set; } = "Pending";
    public string? SourceType { get; set; }
    public string? ReferenceId { get; set; }
    public string? FailureReason { get; set; }
    public DateTime? CompletedAt { get; set; }
    public Reward? Reward { get; set; }
}

File: lib/data/models/reward_transaction.dart
import 'json_helpers.dart';

class RewardTransaction {
  final int transactionId;
  final int reportId;
  final int rewardId;
  final int points;
  final String type;
  final String description;
  final DateTime? createdAt;

  const RewardTransaction({
    this.transactionId = 0,
    this.reportId = 0,
    this.rewardId = 0,
    this.points = 0,
    this.type = '',
    this.description = '',
    this.createdAt,
  });

  factory RewardTransaction.fromJson(Map<String, dynamic> json) {
    return RewardTransaction(
      transactionId: JsonHelpers.intValue(json, [
        'transactionId',
        'TransactionId',
        'id',
        'Id',
      ]),
      reportId: JsonHelpers.intValue(json, ['reportId', 'ReportId']),
      rewardId: JsonHelpers.intValue(json, ['rewardId', 'RewardId']),
      points: JsonHelpers.intValue(json, ['points', 'Points']),
      type: JsonHelpers.stringValue(json, [
        'type',
        'Type',
        'transactionType',
        'TransactionType',
      ]),
      description: JsonHelpers.stringValue(json, [
        'description',
        'Description',
      ]),
      createdAt: JsonHelpers.dateValue(json, ['createdAt', 'CreatedAt']),
    );
  }
}
